### ISSUE TITLE
indent: invalid identation req after multiline sting

### ISSUE THREAD

**Issue by puzrin** [eslint/eslint#5073]
v2.0.0-beta.2

I tried to migrate from https://github.com/nodeca/eslint-plugin-nodeca and found one problem.

The following code:

``` js
/* eslint indent: [2, 2] */

foo.bar()
   .baz(function () {

  var a;
  // function body
});


foo.bar(
    baz,
    baq,
    function () {

  var a;
  // function body
});
```

give this errors:

```
   6:3  error  Expected indentation of 5 space characters but found 2  indent
   8:1  error  Expected indentation of 3 space characters but found 0  indent
  16:3  error  Expected indentation of 6 space characters but found 2  indent
  18:1  error  Expected indentation of 4 space characters but found 0  indent
```

I expect no errors. See real code sample: https://github.com/nodeca/nodeca.core/blob/master/lib/autoload/hooks/init/faye.js#L240-L249
##

**Comment by eslintbot** [comment]
@puzrin Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by gyandeeps** [comment]
Can you plz explain what is the actual output vs your expected output?

Also we have issue open which deals with multiline and starting node issue: https://github.com/eslint/eslint/issues/3737 and

**Comment by puzrin** [comment]
Thanks @gyandeeps, updated description to more comprehensive form.

**Comment by gyandeeps** [comment]
This is a known issue because of the multi-line nature of the arguments. 
So this is how the code is expected by the rule:

```js
/* eslint indent: [2, 2] */

foo.bar().baz(function () {

  var a;
  // function body
});


foo.bar(baz,baq,function () {

  var a;
  // function body
});
```

**Comment by puzrin** [comment]
I understand, but i think that's too strict.

**Comment by gyandeeps** [comment]
Right now but once we have the multi-line support then it will support what you are trying to do. Its a big challenge to incorporate multi-line due the different patterns it can have. We will get their sometime in future.

**Comment by nzakas** [comment]
@gyandeeps so is this a bug? An enhancement?

Also, looks like the indent bugs are piling up. Do you have a time table for addressing them?

**Comment by gyandeeps** [comment]
This is an enhancement which has been mentioned in couple of other accepted issues.
I think we can close this.

I dont have a time table but i can start looking into it.

**Comment by nzakas** [comment]
Can you mention the duplicate issue here and close it?

**Comment by gyandeeps** [comment]
#1801 and #3737

**Comment by puzrin** [comment]
I don't see any duplication in refered issues. Those suggest to enforce indent control. While this ticket is strictly opposite - allow to relax indent control. IMHO it worth to reopen and discuss possible options.

**Comment by puzrin** [comment]
Thanks! I can try to care about this issue in the nearest days.

- will it be accepted?
- could anybody suggest option name (see samples from the first post)?

Short summary  of difference with jshint & eslint-plugin-nodeca/indent:

After some splitted content indentation has no strict requirement. Next block start can be shifted to (indent * XX) spaces. The following lines should be allligned appropriate to the first one of the block. If someone interested in details, i'll dig sources and write exactly all conditions.

**Comment by puzrin** [comment]
Any thoughts? I could try to implement it. The only things i need to know - will it be accepted & option name.

**Comment by ilyavolodin** [comment]
@gyandeeps @BYK could you take a look please? I know you had plans to rewrite indent rule in the future.

**Comment by nzakas** [comment]
Can someone indicate if this is a bug or a enhancement request?

**Comment by gyandeeps** [comment]
Seems like an enchancement but I am not quite sure what changes are being proposed here?
CC @puzrin 
Can you give a more concrete proposal (with examples) of what are you trying to accomplish here?

**Comment by Kirill89** [comment]
This is useful in case of chained call expressions (like in mongoose). I think should be allowed to indent function body from first identifier in chain. Could you take a look to [this example](https://github.com/nodeca/eslint-plugin-nodeca/blob/master/lib/indent.js#L90-L100).

**Comment by puzrin** [comment]
@gyandeeps is Kirill98's explanation enougth?

Suggestion is to allow any positive indent of function's body, counting from the first element of chain. Probably, via additional option. Examples are in the first post.

**Comment by puzrin** [comment]
Sorry, if my question looks stupid, i didn't tracked all changes in eslint core. What is that "multi-line support problem" and why it is related to this issue? As a eslint user, i see 2 things:

- "original" https://www.npmjs.com/package/eslint-plugin-nodeca extension "works as needed" (with relaxed indent control in desired places).
- native (built-in) was based on `eslint-plugin-nodeca`, but has some limitations.

My logic is simple - if it was doable in one plugin, that can be doable in another one.

**Comment by BYK** [comment]
>My logic is simple - if it was doable in one plugin, that can be doable in another one.

Well if you are so sure why don't you try to submit a PR that fixes the problem in the indent rule? ;)

**Comment by ilyavolodin** [comment]
@puzrin Problem is, our implementation of indent rules does a lot more then original `nodeca` implementation. It doesn't make it impossible to add this in, just very hard.

**Comment by puzrin** [comment]
@BYL see https://github.com/eslint/eslint/issues/5073#issuecomment-178776875.

@ilyavolodin i could try to help, if anyone can confirm that it's needed.

**Comment by ilyavolodin** [comment]
@BYK @gyandeeps could you work with @puzrin to figure out how this should be done? I think we would need an option for this with probably three values, something like:
```js
multiline: "no-indent | parent | align"
```
And I think it should be set to "parent" by default. (parent would be indentation of parent + 1 indent level)

**Comment by nzakas** [comment]
Can we coalesce all of the various indent issues where people are suggesting a different type of option format into a single issue with a concrete proposal? I think there's a lot of information spread across multiple issues -- it would be great to have one place where this conversation was happening.

**Comment by BYK** [comment]
I think those three values make sense but I'm not sure if we should put them under the `multiline` key. I've talked about this in a few other issues: I think these should be part of separate configs we do for `VariableDeclaration` etc. For this instance, it would be for `MemberExpression` and/or `BlockStatement.body` (or just `BlockStatement`).

Also 👍 to @nzakas

**Comment by ilyavolodin** [comment]
Looking at this issue again, I think there are three completely different proposals that are currently not supported by indent rule here. 
**First is chain alignment:**
```js
foo.bar()
   .baz(function () {
```
Under normal conditions, baz should be indented 2 spaces, however, it's indented 3 to align with the dot on the previous line.
**Second is argument alignment:**
```js
foo.bar(
    baz,
    baq,
    function () {
```
Same idea here, should be indented 2 spaces, but indent with 3 instead to align with... I'm not exactly sure what this aligning with at all, to be honest, this looks like a really arbitrary style to me. I guess the option here would be not to check indentation of multi-line arguments of a function.
**Three is option to reset alignment of body of function passed in as argument**
```js
foo.bar()
   .baz(function () {

  var a;
  // function body
});
```
This also seems really strange and arbitrary to me. So I would say, again, the only option that I think we might add for this case is to just ignore those completely.

@puzrin Did I summarize this correctly? Would you be fine with just an option to not enforce any indentation for second/third case?

**Comment by Kirill89** [comment]
@ilyavolodin I think you should add an option to not enforce indentation in all listed cases. Next code should be also valid:

```javascript
let result = foo.bar()
              .baz(function () {/*...*/});
```

And I think you shouldn't just skip indentation check. Look at this examples:

First case:

```javascript
foo.bar()
      .baz() // <- bad indentation
   .bat(function () {
```

Second case:

```javascript
foo.bar(
       baz, // <- bad indentation
    baq,
    function () {
```

Third case:

```javascript
foo.bar()
   .baz(function () {

  var a;
     var b; // <- bad indentation
  // function body
});
```

**Comment by kaicataldo** [comment]
@eslint/eslint-team There hasn't been any progress made on this issue in a while. Not sure if we should close because it looks like we might have distilled the indent issues into three fixes based on these two comments:

https://github.com/eslint/eslint/issues/5073#issuecomment-228839126
https://github.com/eslint/eslint/issues/5073#issuecomment-230079291

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get enough support from the team and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after a very long time tend to never do it, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

**Comment by puzrin** [comment]
@not-an-aardvark could you mark such issues with some label? They become lost after such type of closure, without chance to find.

**Comment by not-an-aardvark** [comment]
I'm not sure what you mean. Could you clarify what type of label you would want?

**Comment by puzrin** [comment]
If you mark such issues as "timed-out" (or some better name), it will be easy to navigate those later. Usually, such issues still have good ideas.

I remember one more, closed the same way - it was suggested to show output on the fly, because eslint seems to hang on big projects.
